ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/JmsXAResourceRecoveryHelper.java

```diff
@@ -177,4 +268,11 @@ private void disconnect() throws XAException {
         }
     }
 
+    private void assertIfRecoveryScanWasStarted() throws XAException {
+        if (delegate == null) {
+            LOGGER.warn("XAResource.recover has to be called first with XAResource.TMSTARTRSCAN argument");
+            throw new XAException(XAException.XAER_PROTO);
```

Comment:
Ah ok, that sounds good! Shouldn't I try to close connection in recover(XAResource.TMENDSCAN) too (in addition to commit/rollback/forget)? It is executed in "catch" clause if recover(XAResource.TMSTARTRSCAN) fails.